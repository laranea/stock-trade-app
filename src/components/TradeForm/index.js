// @flow
import React from 'react';
import styled from 'styled-components';

import type { StockSymbols } from '../../types/stockTypes';
import type { TradeSubmission } from '../../types/tradeTypes';
import Widget from '../Widget';

const FormRow = styled.div`
  margin-top: 10px;

  &:first-child {
    margin: 0;
  }
`;

const Input = styled.input`
  width: 100%;
  padding: 0.625rem;
  border-radius: 0.625rem;
  border: 0 none;
  background: #ededed;
`;

const Select = styled.select`
  width: 100%;
  padding: 0.625rem;
  border-radius: 0.625rem;
  border: 0 none;
  background: #ededed;
`;

const Button = styled.button`
  width: 100%;
  padding: 0.625rem;
  border-radius: 0.625rem;
  border: 0 none;
  background: #337ab7;
  color: #ffffff;
  cursor: pointer;

  &[disabled] {
    opacity: 0.5;
  }
`;

type TradeFormProps = {
  handleAddTrade: TradeSubmission => void,
  handleGetStockSymbols: () => void,
  stockSymbols: Array<StockSymbols>
};

type TradeFormState = TradeSubmission & {};

export default class TradeForm extends React.Component<
  TradeFormProps,
  TradeFormState
> {
  state = {
    price: 0,
    quantity: 0,
    stockSymbol: ''
  };

  componentDidMount(): void {
    this.props.handleGetStockSymbols();
  }

  handleInputChange = (event: SyntheticInputEvent<EventTarget>) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  handleSubmitTrade = () => {
    this.props.handleAddTrade({ ...this.state });
    this.setState({
      price: 0,
      quantity: 0,
      stockSymbol: ''
    });
  };

  render() {
    const { quantity, stockSymbol, price } = this.state;

    return (
      <Widget title="Make a trade">
        <form>
          <FormRow>
            <Select
              placeholder="Stock"
              name="stockSymbol"
              onChange={this.handleInputChange}
              value={stockSymbol}
            >
              <option>Choose Stock</option>
              {this.props.stockSymbols.map(symbol => {
                return (
                  <option key={symbol.id} value={symbol.stockSymbol}>
                    {symbol.stockSymbol}
                  </option>
                );
              })}
            </Select>
          </FormRow>
          <FormRow>
            <Input
              placeholder="Price"
              name="price"
              type="number"
              onChange={this.handleInputChange}
              value={price > 0 ? price : ''}
            />
          </FormRow>
          <FormRow>
            <Input
              placeholder="Quantity"
              name="quantity"
              type="number"
              onChange={this.handleInputChange}
              value={quantity > 0 ? quantity : ''}
            />
          </FormRow>
          <FormRow>
            <Button
              type="submit"
              name="submit-button"
              disabled={quantity === 0 || stockSymbol === '' || price === 0}
              onClick={this.handleSubmitTrade}
            >
              Submit trade
            </Button>
          </FormRow>
        </form>
      </Widget>
    );
  }
}
