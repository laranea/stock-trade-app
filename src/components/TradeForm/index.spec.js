import React from 'react';
import { shallow } from 'enzyme';

import TradeForm from './';

const handleAddTradeMock = jest.fn();
const handleGetStockSymbolsMock = jest.fn();
const stockEvent = { target: { name: 'stockSymbol', value: 'TEST' } };
const priceEvent = { target: { name: 'price', value: 20 } };
const quantityEvent = { target: { name: 'quantity', value: 100 } };
const stockSymbolsMock = [{ id: 1, stockSymbol: 'TEST ' }];

describe('TradesForm component', () => {
  it('should match snapshot', () => {
    const component = shallow(
      <TradeForm
        handleAddTrade={handleAddTradeMock}
        handleGetStockSymbols={handleGetStockSymbolsMock}
        stockSymbols={stockSymbolsMock}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should call handleGetStockSymbols on load', () => {
    shallow(
      <TradeForm
        handleAddTrade={handleAddTradeMock}
        handleGetStockSymbols={handleGetStockSymbolsMock}
        stockSymbols={stockSymbolsMock}
      />
    );
    expect(handleGetStockSymbolsMock).toHaveBeenCalled();
  });

  describe('When filling in form', () => {
    let component;

    beforeEach(() => {
      component = shallow(
        <TradeForm
          handleAddTrade={handleAddTradeMock}
          handleGetStockSymbols={handleGetStockSymbolsMock}
          stockSymbols={stockSymbolsMock}
        />
      );
    });

    it('should have all state fields empty by default', () => {
      expect(component.instance().state.stockSymbol).toEqual('');
      expect(component.instance().state.price).toEqual(0);
      expect(component.instance().state.quantity).toEqual(0);
    });

    it('should set stockSymbol field', () => {
      component.find("[name='stockSymbol']").simulate('change', stockEvent);
      expect(component.instance().state.stockSymbol).toEqual(
        stockEvent.target.value
      );
    });

    it('should set price field', () => {
      component.find("[name='price']").simulate('change', priceEvent);
      expect(component.instance().state.price).toEqual(priceEvent.target.value);
    });

    it('should set quantity field', () => {
      component.find("[name='quantity']").simulate('change', quantityEvent);
      expect(component.instance().state.quantity).toEqual(
        quantityEvent.target.value
      );
    });

    describe('When submitting the form', () => {
      it('should be able to submit form with fields completed', () => {
        expect(handleAddTradeMock).not.toHaveBeenCalled();
        component.find("[name='stockSymbol']").simulate('change', stockEvent);
        component.find("[name='price']").simulate('change', priceEvent);
        component.find("[name='quantity']").simulate('change', quantityEvent);
        component.find("[name='submit-button']").simulate('click');
        expect(handleAddTradeMock).toHaveBeenCalled();
        expect(component.instance().state.stockSymbol).toEqual('');
        expect(component.instance().state.price).toEqual(0);
        expect(component.instance().state.quantity).toEqual(0);
      });
    });
  });
});
