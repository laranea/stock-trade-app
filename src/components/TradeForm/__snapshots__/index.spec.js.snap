// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradesForm component should match snapshot 1`] = `
<Widget
  title="Make a trade"
>
  <form>
    <styled.div>
      <styled.select
        name="stockSymbol"
        onChange={[Function]}
        placeholder="Stock"
        value=""
      >
        <option>
          Choose Stock
        </option>
        <option
          key="1"
          value="TEST "
        >
          TEST 
        </option>
      </styled.select>
    </styled.div>
    <styled.div>
      <styled.input
        name="price"
        onChange={[Function]}
        placeholder="Price"
        type="number"
        value=""
      />
    </styled.div>
    <styled.div>
      <styled.input
        name="quantity"
        onChange={[Function]}
        placeholder="Quantity"
        type="number"
        value=""
      />
    </styled.div>
    <styled.div>
      <styled.button
        disabled={true}
        name="submit-button"
        onClick={[Function]}
        type="submit"
      >
        Submit trade
      </styled.button>
    </styled.div>
  </form>
</Widget>
`;
